clients/drcachesim/tools/common/decode_cache.h

```diff
@@ -0,0 +1,390 @@
+     * \p decode_cache_t was constructed with \p persist_decoded_instrs_
+     * set to false, in which case no heap allocation takes place.
+     */
+    virtual void
+    set_decode_info_derived(void *dcontext,
+                            const dynamorio::drmemtrace::_memref_instr_t &memref_instr,
+                            instr_t *instr) = 0;
+
+    bool is_valid_ = false;
+};
+
+/**
+ * Decode info including the full decoded instr_t. This should be used with an
+ * \p decode_cache_t constructed with \p persist_decoded_instrs_ set to
+ * true.
+ */
+class instr_decode_info_t : public decode_info_base_t {
+public:
+    ~instr_decode_info_t();
+    instr_t *
+    get_decoded_instr();
+
+private:
+    void
+    set_decode_info_derived(void *dcontext,
+                            const dynamorio::drmemtrace::_memref_instr_t &memref_instr,
+                            instr_t *instr) override;
+
+    instr_t *instr_ = nullptr;
+    void *dcontext_ = nullptr;
+};
+
+/**
+ * Base class for \p decode_cache_t.
+ *
+ * This is used to allow sharing the static data members among all template instances
+ * of \p decode_cache_t.
+ */
+class decode_cache_base_t {
+protected:
+    /**
+     * Constructor for the base class, intentionally declared as protected so
+     * \p decode_cache_base_t cannot be instantiated directly but only via
+     * a derived class.
+     */
+    decode_cache_base_t() = default;
+    virtual ~decode_cache_base_t();
+
+    static std::mutex module_mapper_mutex_;
+    static std::unique_ptr<module_mapper_t> module_mapper_;
+    // XXX: Maybe the ownership and destruction responsibility for the modfile bytes
+    // should be given to module_mapper_t instead.
+    static char *modfile_bytes_;
+    static int module_mapper_use_count_;
+
+    /**
+     * use_module_mapper_ describes whether we lookup the instr encodings
+     * from the module map, or alternatively from embedded-encodings in the
+     * trace.
+     *
+     * Note that we store our instr encoding lookup strategy as a non-static
+     * data member, unlike module_mapper_t which is static and shared between
+     * all decode_cache_t instances (even of different template types).
+     * Some analysis tools may deliberately want to look at instr encodings
+     * from the module mappings, but that strategy does not provide JIT
+     * encodings which are present only as embedded-encodings in the trace.
+     * In such a case, other concurrently running analysis tools should still
+     * be able to see encodings for JIT code.
+     */
+    bool use_module_mapper_ = 0;
+
+    /**
+     * Initializes the module_mapper_ object using make_module_mapper() and performs
+     * other bookkeeping and prerequisites.
+     */
+    std::string
+    init_module_mapper(const std::string &module_file_path,
+                       const std::string &alt_module_dir);
+
+    /**
+     * Returns the address where the encoding for the instruction at trace_pc can
```

Comment:
Wait how is the address a std::string?

The output is the 2nd param right?  Return value is error?  Please reword and expand the comment.

Reply:
Yes. Reworded.